Client/SafeBrowsing/GSafeBrowsing.swift

```diff
@@ -0,0 +1,316 @@
+
+import Foundation
+import CommonCrypto
+import Shared
+
+private let log = Logger.browserLogger
+
+class SafeBrowsingClient {
+    private static let apiKey = "DUMMY_KEY"
+    private static let maxBandwidth = 2048
+    private static let maxDiskSpace = 4096
+    private static let clientId = AppInfo.baseBundleIdentifier
+    private static let version = AppInfo.appVersion
+    
+    private let baseURL = "https://safebrowsing.brave.com"
+    private let session = URLSession(configuration: .ephemeral)
+    private let database = SafeBrowsingDatabase()
+    
+    public static let shared = SafeBrowsingClient()
+    
+    private init() {
+    }
+    
+    func find(_ hashes: [String], _ completion: @escaping (_ isSafe: Bool, Error?) -> Void) {
+        let discoveredHashes = database.find(hashes)
+        if discoveredHashes.isEmpty {
+            return completion(true, nil)
+        }
+        
+        let clientInfo = ClientInfo(clientId: SafeBrowsingClient.clientId,
+                                    clientVersion: SafeBrowsingClient.version)
+        
+        let threatTypes: [ThreatType] = [.malware,
+                                         .socialEngineering,
+                                         .unwantedSoftware,
+                                         .potentiallyHarmfulApplication]
+        
+        let platformTypes: [PlatformType] = [.any]
+        let threatEntryTypes: [ThreatEntryType] = [.url, .exe]
+        
+        let threatInfo = ThreatInfo(threatTypes: threatTypes,
+                                    platformTypes: platformTypes,
+                                    threatEntryTypes: threatEntryTypes,
+                                    threatEntries: discoveredHashes.map {
+                                        return ThreatEntry(hash: $0, url: nil, digest: nil)
+                                    }
+        )
+        
+        do {
+            let body = FindRequest(client: clientInfo, threatInfo: threatInfo)
+            let request = try encode(.post, endpoint: .fullHashes, body: body)
+            executeRequest(request, type: FindResponse.self) { response, error in
+                if let error = error {
+                    return completion(false, error)
+                }
+                
+                if let response = response {
+                    if !response.matches.isEmpty {
+                        return completion(false, nil)
+                    }
+                }
+                
+                completion(true, nil)
+            }
+        } catch {
+            completion(false, error)
+        }
+    }
+    
+    func fetch(_ completion: @escaping (Error?) -> Void) {
+        let clientInfo = ClientInfo(clientId: SafeBrowsingClient.clientId,
+                                    clientVersion: SafeBrowsingClient.version)
+        
+        let constraints = Constraints(maxUpdateEntries: UInt32(SafeBrowsingClient.maxBandwidth),
```

Comment:
I don't think wee should cap the number of entries? Adding a cap to the db size makes sense, I think we can increase it to 1mb

Reply:
I agree, let's not cap the number of entries, it's not that much bandwidth and limiting this could significantly reduce the protection our users get out of Safe Browsing.

Reply:
Hey guys.. I have been testing this for the last 17 hours.. I CANNOT put no cap.
The variables represent the "number of entries" and not the number of bytes. When I have to calculate the checksum on the database, it requires actually calculating the hash of ALL the entries in the database and comparing against the original checksum that was received from a previous call..

Having to pull in 100's of thousands of entries in at once and then looping all of them kills performance drastically. The app takes over 10 seconds to LAUNCH even when I added massive performance improvements.

Reply:
> > 1. It means I don't have expiration handling yet. I only have it redownloading the DB on every launch. I couldn't implement all of it at once because expiration handling isn't just about clearing the DB. It also has an "in-memory" cache of recently looked up URLs so that if the user refreshes the same page, it doesn't do a DB lookup and doesn't require resorting the entire DB upon insertion. Also, most of the above code is ported from Go-Lang: https://github.com/google/safebrowsing
> 
> Ok, thanks for the clarification. This is fine for a first PR. I can understand the desire to land this in stages because it's a lot of code. We'll just need to make sure that it cannot be enabled by users (compile time flag?) when we land this because we're not allowed to use the Google service without all of the caching in place.
> 
> > 1. Line 54, 260, 264 of GSafeBrowsingDatabase.swift.
> > 2. I supported variable length with optimization for 4-byte prefixes: Line 31 of GSafeBrowsingDatabase.swift
> 
> Thanks for the pointers!
> 
> There's one more thing I couldn't easily see in the PR: when is the update scheduled at browser startup? Is it randomly done [between 0 and 1 minutes from browser startup](https://developers.google.com/safe-browsing/v4/request-frequency#update-requests)?

I did NOT do the "random" initialization between 0 and 1 minute.. I immediately initialize the SB-Database upon launch. Not sure what the logic is behind doing it at random.. That gives more chance for the user to run into a bad page.. IE: If the threats are fetched at 45s, that means for 45s, the user is without protection?

I have no problem doing it at random but it doesn't seem like the greatest security..

For the `scheduled` updates, I have it implemented at: Line 25 & 162 of GSafeBrowsing.swift
It will trigger an update automatically when the database becomes stale.

Reply:
> Not sure what the logic is behind doing it at random.

That condition is there to prevent server overload. I imagine the reasoning is something like: if a lot of people wake up on the East coast at 8am and check the news in their browser, then a lot of people could hit the update server at once.

You're right that it is a trade-off between having protection earlier v. server scalability. In this case, Google has decided what the "right" answer is and that's to protect their server infrastructure. There are a lot of things in Safe Browsing where service availability is prioritized over security. It does make sense in a way too because if the service were to be overloaded and slow to respond (or not responding) then that would also hurt security since when you can't get an answer from the server, you don't show a warning.

Reply:
> Hey guys.. I have been testing this with different solutions for the last day or so.. We need to have a cap.

Do we need both limits for perf reasons or just the limit on the number of entries?

The more the better, but obviously if there are perf constraints, then we have to do something reasonable. Waiting 10 seconds for the list to be downloaded after opening the browser is not exactly a good user experience as you point out.

Reply:
Ok I added the code to execute it at random between 0 and 1 minute.

For the performance, no we don't need to limit "both". We can limit one if we like but the limit (as you said), should be reasonable.

We won't have to limit `bandwidth` if we want but we should limit the local db entries because searching would take longer as it grows. For bandwidth, the initial `update` call might take long.

This was the reason I limited both. Fast initial fetch and fast lookup. Everything else goes through the server `find` call and gets cached.